html/browsers/browsing-the-web/back-forward-cache/resources/helper.sub.js

```diff
@@ -0,0 +1,129 @@
+// Helpers called on the main test HTMLs.
+// Functions in `RemoteContext.execute_script()`'s 1st argument are evaluated
+// on the executors (`executor.html`), and helpers available on the executors
+// are defined in `executor.html`.
+
+const originSameOrigin =
+  location.protocol === 'http:' ?
+  'http://{{host}}:{{ports[http][0]}}' :
+  'https://{{host}}:{{ports[https][0]}}';
+const originSameSite =
+  location.protocol === 'http:' ?
+  'http://{{host}}:{{ports[http][1]}}' :
+  'https://{{host}}:{{ports[https][1]}}';
+const originCrossSite =
+  location.protocol === 'http:' ?
+  'http://{{hosts[alt][www]}}:{{ports[http][0]}}' :
+  'https://{{hosts[alt][www]}}:{{ports[https][0]}}';
+
+const executorPath =
+  '/html/browsers/browsing-the-web/back-forward-cache/resources/executor.html?uuid=';
+
+// Asserts that the executor `target` is (or isn't, respectively)
+// restored from BFCache. These should be used in the following fashion:
+// 1. Call prepareNavigation() on the executor `target`.
+// 2. Navigate the executor to another page.
+// 3. Navigate back to the executor `target`.
+// 4. Call assert_bfcached() or assert_not_bfcached() on the main test HTML.
+async function assert_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status === 'BFCached', 'Should be BFCached');
+}
+
+async function assert_not_bfcached(target) {
+  const status = await getBFCachedStatus(target);
+  assert_implements_optional(status !== 'BFCached', 'Should not be BFCached');
+}
+
+async function getBFCachedStatus(target) {
+  const [loadCount, isPageshowFired] =
+    await target.execute_script(() => [window.loadCount, window.isPageshowFired]);
+  if (loadCount === 1 && isPageshowFired === true) {
+    return 'BFCached';
+  } else if (loadCount === 2 && isPageshowFired === false) {
+    return 'Not BFCached';
+  } else {
+    // This can occur for example when this is called before first navigating
+    // away (loadCount = 1, isPageshowFired = false), e.g. when
+    // 1. sending a script for navigation and then
+    // 2. calling getBFCachedStatus() without waiting for the completion of
+    //    the script on the `target` page.
+    assert_unreached(
+      `Got unexpected BFCache status: loadCount = ${loadCount}, ` +
+      `isPageshowFired = ${isPageshowFired}`);
+  }
+}
+
+// Always call `await remoteContext.execute_script(waitForPageShow);` after
+// triggering to navigation to the page, to wait for pageshow event on the
+// remote context.
+const waitForPageShow = async () => {
+  while (!window.pageShowPromise) {
+    await new Promise(resolve => setTimeout(resolve, 50));
+  }
```

Comment:
I would like to remove the exception from `lint.ignore `and the setTimeout. I can't really understand why it is needed. Coud you please explain it to me?

I think I would just remove this function and replace occurences of:
```js
await pageA.execute_script(waitForPageShow);
```
by
```js
await pageA.execute_script(() => window.pageShowPromise);
```
and remove the exception from `lint.ignore`.

Reply:
This `setTimeout` came from `waitForPageShow` for jgraham's version of `execute_script`, where the script can be evaluated even before `pageShowPromise` is set, and thus I have to wait for `pageShowPromise` to be set via `setTimeout`.
Now this PR is based on `common/dispatcher` where I can ensure `pageShowPromise` is set before the injected script evalution, so I removed the `setTimeout` and the exception in `lint.ignore`.

`waitForPageShow` is still kept, because it is introduced to decouple the details of `execute_script` impl from test body. When we change the underlying `execute_script` impl (e.g. switching to jgraham's proposal), `waitForPageShow` might be changed (e.g. reverted to the `setTimeout` version), but test bodies can remain unmodified. Replacing this to `() => window.pageShowPromise` removes this benefit.

Reply:
Removing the SetTimeout for now and keeping waitForPageShow for now like you did sounds good to me. Thanks!